CLoadTags fixes

nwf opened the pull request:
Both RISC-V and MIPS had hard-coded the stride to 8, which was missed when
CAP_TAG_GET_MANY_SHFT was recently changed.

MIPS failed to check that the entire region was within the auth cap bounds

nwf-msr committed:
CLoadTags fixes

Both RISC-V and MIPS had hard-coded the stride to 8, which was missed when
CAP_TAG_GET_MANY_SHFT was recently changed.

Move these to cheri-common and move the implementation to cap_check_common_reg.

arichardson reviewed (commented):
These look almost identical, should we just moved them to cheri-common? Or is the MIPS behaviour difference important?

nwf commented:
Moving them to `cheri-common` seems fine to me.

LawrenceEsswood commented:
The checks here are basically cap_check_common_reg again as far as I can see. The only difference seems to be behaviour around CAP_PERM_LOAD_CAP. Can we just add a 'CAP_PERM_LOAD_CAP' check to the big if...else block in cap_check_common_reg, then create an perms_for_load_tags() that returns (CAP_PERM_LOAD | CAP_PERM_LOAD_CAP) on RISCV/MIPS and CAP_PERM_LOAD on morello? Then all platforms could use the same loadtags logic, and we wouldn't have all this duplicate logic.

nwf-msr committed:
cheri-common/cap_check_common_reg: LOAD_CAP check

Optionally check for CAP_PERM_LOAD_CAP as well.  This is (now) very rare in
CHERI architectures in general, as we've come to prefer tag clearing.  But for
CLoadTags we've decided that missing LOAD_CAP is such a foot gun that it's
probably worth trapping.

As cap_check_common_reg is marked as QEMU_ALWAYS_INLINE, this should have little
impact on anything on the fast path, for which required_perms is a constant.

nwf commented:
Test failure looks unrelated?

arichardson commented:
> Test failure looks unrelated?

Yes that's just the builder being too busy. We should probably just skip that test since it doesn't make much sense anyway.

arichardson commented:
> Test failure looks unrelated?

Fixed in https://github.com/CTSRD-CHERI/cheritest/commit/e7f15d4420ea8635297783aa5994c915cf0049c2, will re-trigger CI
